postgresql-simple/tests/tests.hs

```diff
@@ -0,0 +1,25 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Main where
+
+import Control.Exception
+import Test.Tasty
+import Database.PostgreSQL.Simple
+import qualified Database.YeshQL.PostgreSQL.Tests as Tests
+
+main :: IO ()
+main = withConn (defaultMain . allTests)
+  where
+    allTests conn = testGroup "All Tests" (Tests.tests conn)
+
+withConn :: (Connection -> IO a) -> IO a
+withConn f =
+  bracket (connectPostgreSQL "host='localhost' user='postgres'") close $ \c ->
```

Comment:
This is bad; tests should not assume an installed postgresql server, and they should definitely not touch any actual postgresql clusters. I would prefer using a mock database, similar to how the HDBC backend does it; just verify that the functions use the right queries and send the right data, and that the data that comes back out is converted correctly.

If you really really want to test against a real postgresql database, make it opt-in, and configurable, so that users can decide whether they want the test suite to touch their cluster, and if so, how.

Reply:
> This is bad; tests should not assume an installed postgresql server, and they should definitely not touch any actual postgresql clusters. I would prefer using a mock database, similar to how the HDBC backend does it; just verify that the functions use the right queries and send the right data, and that the data that comes back out is converted correctly.

Note that the tests run in a transaction which is rolled back after the tests so they are relatively “harmless”. I also think testing against the actual DB is useful since in the end that’s what users care about.

That said I agree that it would be nice to be able to run the tests without a running postgresql server (you are going to need to install `libpq` since `postgresql-simple` links against that). However, contrary to `HDBC` `postgresql-simple` doesn’t abstract over the connection and the connection is also fixed in the type classes used for encoding/decoding to SQL values so using a mock database would require coming up with a fair amount of custom abstractions just for the purposes of testing and I’m not sure that’s really worth the effort. These abstractions are also going to leak into the type signatures that users see or we make things configurable which adds even more complexity.

Reply:
Hmm, I see... can we then maybe make it such that the specific connection used for the tests can be configured using environment variables, and that it fails in a useful way when the variables aren't set?

Reply:
I’ve changed it such that the connection string is read from the `YESHQL_POSTGRESQL_CONNSTRING` environment variable and if it is not set it bails out and suggests setting it.